### PR TITLE
feat: offer to restart workspace when ttl is changed

### LINKED ISSUE #5080: bug: "Stops at" UI element does not reflect the set auto-stop value
### What happened?

I'm running on `0.12.7`. after setting my workspace auto-stop to 5 hours, the `STOPS AT` value did not change and continued to show 1 hour from the current time.

Screen Shot 2022-11-14 at 9 49 50 PM

Screen Shot 2022-11-14 at 9 50 05 PM

### What did you expect to happen?

I expected my workspace `STOPS AT` value to be 5 hours from the current time.

### PR DESCRIPTION
**Pull request by presleyp**
Closes #5080 

If ttl is changed, a dialog will show asking if you want the workspace to restart. If you choose yes, you will go to the workspace page and then see the restart process. This results in the new ttl taking effect.

### CODE REVIEW

File: site/src/pages/WorkspaceSchedulePage/WorkspaceSchedulePage.tsx
Comment (Kira-Pilot) [inline site/src/pages/WorkspaceSchedulePage/WorkspaceSchedulePage.tsx:28]: Can we make translations for these?

File: site/src/pages/WorkspaceSchedulePage/WorkspaceSchedulePage.tsx
Comment (Kira-Pilot) [inline site/src/pages/WorkspaceSchedulePage/WorkspaceSchedulePage.tsx:147]: I'm surprised to see us passing this as part of `Location`. I would only expect to see routing-specific variables in that object.

Is there a reason we can't post a new build for the workspace from inside of the workspaceSchedule machine after the TTL post returns, then navigate to the workspace page?

File: site/src/pages/WorkspacePage/WorkspaceReadyPage.tsx
Comment (Kira-Pilot) [inline site/src/pages/WorkspacePage/WorkspaceReadyPage.tsx:77]: nice comment!

File: site/src/pages/WorkspaceSchedulePage/WorkspaceSchedulePage.tsx
Comment (presleyp) [reply site/src/pages/WorkspaceSchedulePage/WorkspaceSchedulePage.tsx:147]: Since starting a workspace takes several seconds, and since the workspace page already handles errors for it, I wanted to go back to the workspace page first and then make the api call. But since our workspace X service is only alive when the page is mounted (which we could change, but I was afraid that would be too big a change), I can't send it a message from the schedule page. Maybe I could fire and forget from the schedule page since the workspace page gets updates pushed from the server?

Comment (Kira-Pilot) [reply site/src/pages/WorkspaceSchedulePage/WorkspaceSchedulePage.tsx:147]: Yeah, good idea: I think you could fire and forget! I think the only waiting we need to do is ensuring that TTL post finishes first, so we don't restart the workspace if something goes awry.

File: site/src/pages/WorkspaceSchedulePage/WorkspaceSchedulePage.test.tsx
Comment (Kira-Pilot) [inline site/src/pages/WorkspaceSchedulePage/WorkspaceSchedulePage.test.tsx:21]:
```suggestion
} from "components/WorkspaceScheduleForm/WorkspaceScheduleForm"
```

File: site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts
Comment (Kira-Pilot) [inline site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts:152]: This shouldn't be "autoStopChanged"?

Comment (presleyp) [reply site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts:152]: Wow, it effectively is but I named it wrong. Good catch!

File: site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts
Comment (Kira-Pilot) [inline site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts:178]: Is this a 'finalState'? Should we label it as such? https://xstate.js.org/docs/guides/final.html#api

File: site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts
Comment (Kira-Pilot) [inline site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts:274]: What happens if this errors? I don't see a catch, so I'm not sure how we'd enter the error state on line 157 of this machine. I would want to avoid prompting users to restart their workspace if no change was actually made.

Comment (presleyp) [reply site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts:276]: I wasn't sure if it was a problem that I'm not returning the result from both promises here, so I tried hard-coding each api call to return a rejected promise, and it behaved the way I intended: if one api call errors and that part of the schedule is not changed, there's no error. If the error api call is called, you stay on the form and get an error message. If you change the start but not the stop, you don't see the prompt.

File: site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts
Comment (BrunoQuaresma) [inline site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts:102]: Permissions are loaded in the initial app loading in the auth machine so I don't think we need to load it again. If there are new permissions, I would add them there.

I would do is something like this:
```
const permissions = usePermissions()
const [workspaceScheduleState] = useMachine(workspaceSchedule, { context: { permissions } })
```

File: site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts
Comment (BrunoQuaresma) [inline site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts:136]: Maybe, we could condense the `gettingPermissions`, `gettingWorkspace`, and `gettingTemplate` in a single state called `loading` + a `load` service that returns all this info. I'm thinking we are over-creating states making the machine more complex to understand without a clear value. By over-creating, I mean states that are not reflected in the UI/UX.

**Review by BrunoQuaresma — APPROVED** [review]
LGTM. I left a few comments but any of them are blockers. You can just skip them if they don't make sense or make it later, your call!

File: site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts
Comment (presleyp) [reply site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts:102]: Okay, I didn't add the permissions stuff in this PR (the machine just got reformatted), but I'll make a note to look into that later.

File: site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts
Comment (presleyp) [reply site/src/xServices/workspaceSchedule/workspaceScheduleXService.ts:136]: Yeah, on one hand having separate states is helpful for debugging, but on the other hand it's extra states that aren't strictly necessary. We could change the way we handle errors to save whichever error we get going through the whole loading sequence. I'm not going to hold up this PR over it but it's worth considering in the future.

**Pull request merged**
